### PR TITLE
GC: wasmparser: Subtypes

### PR DESCRIPTION
**Pull request by imikushin**
Introducing the `SubType` struct:
```rust
pub struct SubType {
    pub is_final: bool,
    pub supertype_idxs: Vec<u32>,
    pub structural_type: StructuralType,
}
```

Here, `StructuralType` is an enum (previously named `Type`).

`supertype_idxs` is a vector of indexes of supertypes, limited to at most one supertype as per GC MVP.

A valid subtype matches its supertypes. Supertypes cannot be final.

### CODE REVIEW

**Comment by imikushin** [comment]
x86_64-linux and x86_64-windows CI jobs were canceled for some reason.

File: crates/wasmparser/src/readers/core/types.rs
Comment (fitzgen) [inline crates/wasmparser/src/readers/core/types.rs:1034]: Nitpick: could we name this something other than `op_code` here and below in the `FromReader` impl for `SubType`? That makes me think of instruction opcodes where as this doesn't have anything to do with instructions, just types.

Perhaps `type_discriminant`? Or `leading_byte`?

File: crates/wasmparser/src/validator/component.rs
Comment (fitzgen) [inline crates/wasmparser/src/validator/component.rs:960]: Looking at all these conversions between `Type`, `SubType`, `StructuralType`, and each structural type variant, it seems like we could have some more `From` implementations to make this less noisy and more automatic.

To really benefit from this, and because `From` is not transitive by default, I think we would want all of these impls:

```rust
impl From<FuncType> for StructuralType { ... }
impl From<ArrayType> for StructuralType { ... }
impl From<StructType> for StructuralType { ... }

impl From<StructuralType> for SubType { ... }
impl From<FuncType> for SubType { ... }
impl From<ArrayType> for SubType { ... }
impl From<StructType> for SubType { ... }

impl From<SubType> for Type { ... }
impl From<StructuralType> for Type { ... }
impl From<FuncType> for Type { ... }
impl From<ArrayType> for Type { ... }
impl From<StructType> for Type { ... }
```

This can be done in a follow up PR if you prefer.

File: crates/wasmparser/src/validator/core.rs
Comment (fitzgen) [inline crates/wasmparser/src/validator/core.rs:528]:
```suggestion
            // Check the supertype exists, is not final, and the subtype matches it.
```

File: crates/wasmparser/src/validator/types.rs
Comment (fitzgen) [inline crates/wasmparser/src/validator/types.rs:212]: You can use `..` to ignore multiple fields at a time, instead of writing them all out here and below:

```suggestion
            Type::Sub(SubType {
                structural_type: StructuralType::Func(ft),
                ..
            }) => Some(ft),
```

File: tests/local/gc/gc-subtypes.wat
Comment (fitzgen) [inline tests/local/gc/gc-subtypes.wat:7]: I think we need more tests for subtyping, basically every branch of all the `Matches` implementations should be covered by our tests:

* [ ] val types that are and are not the same
* [ ] ref types and each case of same/different nullability
* [ ] heap types that are the same, not the same but sub types, and not subtypes
* [ ] each of the "known"/abstract heap types (`struct`, `noextern`, `eq`, etc)
* [ ] basically, each case in `Matches for HeapType`
* [ ] final vs not final sub types in all combinations
* [ ] super types that the sub type does vs doesn't match
* [ ] too many super types

there are a couple of these partially covered in the `gc-subtypes-invalid.wast` test, but we need to be very thorough and deliberate around exercising all these corners and combinations.

**Review by fitzgen — COMMENTED** [review]
Thanks!

This is looking good, but I have a few nitpicks and also a request for more complete testing.

Once those things are addressed we can merge this.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:1034]: Would `opcode` (without the `_`) work? Seems to be the term used for this purpose in the spec (for example, in https://webassembly.github.io/spec/core/appendix/index-types.html)

Alternatively, how about `type_code`?

File: crates/wasmparser/src/validator/component.rs
Comment (imikushin) [reply crates/wasmparser/src/validator/component.rs:960]: Added an issue to follow this up: #1071

File: crates/wasmparser/src/readers/core/types.rs
Comment (fitzgen) [reply crates/wasmparser/src/readers/core/types.rs:1034]: Ah, I didn't realize the spec used `opcode` for this. In that case, yes, let's go with that!

File: crates/wasm-encoder/src/core/types.rs
Comment (imikushin) [reply crates/wasm-encoder/src/core/types.rs]: Fixed.

Comment (imikushin) [reply crates/wasm-encoder/src/core/types.rs]: Took approach (1).

File: tests/local/gc/gc-subtypes.wat
Comment (imikushin) [reply tests/local/gc/gc-subtypes.wat]: Added more tests.

**Review by fitzgen — APPROVED** [review]
LGTM with a couple minor nitpicks. I'll be on vacation for a week starting tomorrow, so maybe @alexcrichton can merge this once you've addressed the nitpicks.

Thanks!

File: crates/wasm-encoder/src/core/types.rs
Comment (imikushin) [reply crates/wasm-encoder/src/core/types.rs:345]: Reverting this suggestion for now: this seems to have broken the tests locally 🤔

File: tests/local/gc/gc-subtypes.wat
Comment (imikushin) [reply tests/local/gc/gc-subtypes.wat:9]: Already testing these in `gc-subtypes-invalid.wast`. Removing.

File: crates/wasm-encoder/src/core/types.rs
Comment (imikushin) [reply crates/wasm-encoder/src/core/types.rs:332]: Thanks! Great catch.

File: crates/wit-component/src/gc.rs
Comment (imikushin) [reply crates/wit-component/src/gc.rs:583]: I agree, this _is_ kinda ugly. Thanks for pointing it out!

File: crates/wasm-encoder/src/core/types.rs
Comment (imikushin) [reply crates/wasm-encoder/src/core/types.rs]: Using a different approach: turn the option into a vector (and encoding it).

File: crates/wit-component/src/gc.rs
Comment (imikushin) [reply crates/wit-component/src/gc.rs]: Fixed.

**Comment by imikushin** [comment]
@alexcrichton I believe I've addressed all @fitzgen's feedback (thank you!).

File: crates/wast/src/names.rs
Comment (alexcrichton) [inline crates/wast/src/names.rs:54]: I think it would be best to leave out this change since, as you commented on the original issue, this isn't a complete fix. I find that this makes the state of things more confusing since it appears to sometimes work but still sometimes not work.

File: crates/wasmparser/src/validator/core.rs
Comment (alexcrichton) [inline crates/wasmparser/src/validator/core.rs:521]: I think this may need a feature check for `gc`?

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:59]: One thing I might recommend is to remove the `Result` here since indices at this point should always be valid because the definitions of types are validated to all be correct. Additionally it might be better to pass a concrete type here rather than a generic `Fn(u32) -> &SubType` but instead have something like `cx: &Validator` or similar.

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:541]: Could parentheses be used to clarify the order of operations here? I don't know the binding precedence of `||` vs `&&` myself

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:675]: I've mentioned this before, but personally I'm not a fan of catch-all clauses like this. This is very easy to forget to update if a new variant is added because of `#[derive(PartialEq)]` which can lead to further bugs in behavior. Where possible I'd recommend exhaustively matching remaining cases, for example something like:


```
(a @ (HeapType::T1 | HeapType::T2), b) => Ok(a == b),
```

which explicitly documents the list of "catch-all" types being caught here and it's obvious there's no payload so equality can only be identity-based. That way if new types are added in the future this'll generate a compiler error with a non-exhaustive check.

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:763]: I'm not familiar with the GC proposal myself, but a lot of these boolean checks feel iffy to me. For example if the other type is `final`, shouldn't an exact structural match be performed? This seems like it rejects all matching entirely at that point. That seems to me like two types which have the same internal indices may just happen to pass some equality test prior to this but it seems like `(type $t1 (func))` and `(type $t2 (func))` should return "true" for a matches predicate if they're both final. 

I could very well be missing something about the gc proposal, though, since I am not up to speed on it.

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:808]: Like above I'd recommend removing catch-all cases like this, and instead here you can do something like:

```
(StructuralType::Func(a), StructuralType::Func(b)) => a.matches(b, type_at),
(StructuralType::Func(_), _) => false,
```

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:891]: Could you expand this comment? For example this switches the order of the `matches` predicate which seems important to call out here, along with its implications.

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:932]: Another catch-all I think would be good to remove

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [inline crates/wasmparser/src/readers/core/types.rs:1089]: There's a fair amount of duplication in this method, but could that perhaps be removed with something like:

```rust
let is_final = match reader.peek()? {
    0x50 | 0x4e => { /* advance reader */ }
    _ => None,
};
let supertype_idx = if is_final.is_some() {
    // read the supertype
} else {
    None 
};
let structural_type = /* preexisting structural type reading */;

SubType { is_final, supertype_idx, structural_type }
```

File: crates/wasm-mutate/src/info.rs
Comment (alexcrichton) [inline crates/wasm-mutate/src/info.rs:96]: I think it would be best to update `wasm-mutate` a bit differently in this commit. Currently `wasm-mutate` doesn't have support for GC, and that's ok, but if given a binary with gc with these changes it runs a risk of "accepting" the binary and making it invalid. For example this is ignoring all fields other than `structural_type`. 

I think it would be better to keep wasm-mutate as-is where it explicitly rejects gc-using binaries. For wasm-mutate that would look like handling the full `SubType` structure but returning errors if any fields are set that look like gc.

File: crates/wast/src/names.rs
Comment (imikushin) [reply crates/wast/src/names.rs:54]: Reverting. Will start a new PR with a proper fix.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:59]: > One thing I might recommend is to remove the Result here since indices at this point should always be valid because the definitions of types are validated to all be correct.

Actually, the primary use case for `matches()` is to validate if the type definitions are correct.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:541]: Sure thing!

And I thought people didn't like parentheses. Lisp is the best.

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [reply crates/wasmparser/src/readers/core/types.rs:59]: I don't think that's the case? Whenever a type definition is reached as part of validation there are no type-matching queries performed at that time and instead the structure of the defined type is validated. This means that the indices are all valid for the type section after it's read, so in the future when the matching predicate comes into play during the code section for example there are no invalid indices.

I believe this is `Module::add_type` which is where indices are validated and afterwards indices don't need validation.

Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:59]: > I believe this is Module::add_type which is where indices are validated and afterwards indices don't need validation.

Yeah, that's exactly where `matches()` gets called:

```
<wasmparser::readers::core::types::SubType as wasmparser::readers::core::types::Matches>::matches types.rs:763
wasmparser::validator::core::Module::check_subtype core.rs:525
wasmparser::validator::core::Module::add_type core.rs:507
```

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:675]: > because of `#[derive(PartialEq)]`

Not sure I understand what bad things can happen because of it. Everything else seems clear.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:763]: Would it be more clear if we rename `matches` to `derives`?

We cannot derive a subtype from a type that is final. I'm also kinda worried that `matches` has broader meaning, and is potentially a bad naming decision.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:891]: Adding a reference to the GC spec.

File: crates/wasmparser/src/validator/core.rs
Comment (imikushin) [reply crates/wasmparser/src/validator/core.rs:527]: Added.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs]: Removed.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs]: Removed.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:776]: Renamed.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs:1105]: Fixed.

File: crates/wasm-mutate/src/info.rs
Comment (imikushin) [reply crates/wasm-mutate/src/info.rs]: Fixed.

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs]: Also, updated the code according to your comment.

File: crates/wasmparser/src/readers/core/types.rs
Comment (alexcrichton) [reply crates/wasmparser/src/readers/core/types.rs:59]: Ah ok I see that now, but I still think that this `Result` isn't necessary. For example if `check_structural_type` is moved above the subtype check, then everything is guaranteed to be valid while performing a subtype check?

Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs]: Thanks! Updated.

**Comment by imikushin** [comment]
@alexcrichton I believe I've addressed all your comments (so far).

File: crates/wasmparser/src/readers/core/types.rs
Comment (imikushin) [reply crates/wasmparser/src/readers/core/types.rs]: Renamed to `inherits` which should reflect the meaning even better (than `derives`).

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**
